File: Unity/Assets/Eruru.MVVM/Scripts/Shared/MVVMRelayCommand.cs
using System;

namespace Eruru.MVVM {

	public class MVVMRelayCommand {

		readonly Predicate<object> _CanExecute;
		readonly Action<object> _Execute;

		public MVVMRelayCommand (Action<object> execute, Predicate<object> canExecute) {
			_Execute = execute;
			_CanExecute = canExecute;
		}
		public MVVMRelayCommand (Action<object> execute) {
			_Execute = execute;
		}

		public bool CanExecute (object parameter) {
			return _CanExecute == null || _CanExecute (parameter);
		}

		public void Execute (object parameter) {
			if (_Execute != null) {
				_Execute (parameter);
			}
		}

	}

}

File: IntelliJ IDEA/Java/src/com/eruru/mvvm/MVVMRelayCommand.java
package com.eruru.mvvm;

public class MVVMRelayCommand {

	MVVMPredicate1<Object> _CanExecute;
	MVVMAction1<Object> _Execute;

	public MVVMRelayCommand (MVVMAction1<Object> execute, MVVMPredicate1<Object> canExecute) {
		_Execute = execute;
		_CanExecute = canExecute;
	}

	public MVVMRelayCommand (MVVMAction1<Object> execute) {
		_Execute = execute;
	}

	public boolean canExecute (Object parameter) {
		return _CanExecute == null || _CanExecute.invoke (parameter);
	}

	public void execute (Object parameter) {
		if (_Execute != null) {
			_Execute.invoke (parameter);
		}
	}

}
